### PR TITLE
Drop ruby 2.1/2.2/2.3 support

### LINKED ISSUE #2380: Drop older ruby 2.1/2.2/2.3 support
Fluentd v1 now works on ruby 2.1 or later but it has several drawbacks.

- ruby 2.1/2.2 are dead version. ruby team doesn't maintain it.
- Need dig_rb gem for `dig` method: See https://github.com/fluent/fluentd/pull/2372

Removing 2.1/2.2 is good for the further development. td-agent and docker images use ruby 2.3 or later so it doesn't break popular environment.
However removing support without the transition period may break user environment. So support 2.1/2.2 until the end of 2019 is better.

UPDATE: Drop 2.3 together. See https://www.fluentd.org/blog/drop-schedule-announcement-in-2019

### PR DESCRIPTION
**Pull request by repeatedly**
Signed-off-by: Masahiro Nakagawa <repeatedly@gmail.com>

**Which issue(s) this PR fixes**: 
Fixes #2380 
See also: https://www.fluentd.org/blog/drop-schedule-announcement-in-2019

This is for fluentd v1.9.0 which will be released in 2020

**What this PR does / why we need it**: 
Drop older rubies and related gems.

**Docs Changes**:
Need to update supported version

**Release Note**: 
Same as title.

### CODE REVIEW

**Comment by ganmacs** [comment]
ref https://github.com/fluent/fluentd/pull/2372

File: Gemfile [hunk 1]
```diff
@@ -2,10 +2,7 @@ source 'https://rubygems.org/'
 
 gemspec
 
-# https://github.com/socketry/async-io/blob/v1.23.1/async-io.gemspec#L21
-if Gem::Version.create(RUBY_VERSION) >= Gem::Version.create('2.3.0')
-  gem 'async-http', '~> 0.42'
-end
+gem 'async-http', '~> 0.42'
```
Comment (ganmacs) [inline Gemfile]: this gem should be moved into fluentd.gemspec

File: lib/fluent/plugin_helper/http_server.rb [hunk 2]
```diff
@@ -15,7 +15,6 @@
 #
 
 begin
-  # raise if RUBY_VERSION < 2.3.x. see Gemfile
```
Comment (ganmacs) [inline lib/fluent/plugin_helper/http_server.rb]: `rescue` expression is also not necessary ( also Fluent::PluginHelper::HttpServer::Compat::Server is )

File: Gemfile (hunk 1 above)
Comment (repeatedly) [reply Gemfile]: I think fluentd should work with minimal dependencies. So make async-http optional seems good for me. If we can replace cool.io with async gems, async-http should be default dependency.

Comment (ganmacs) [reply Gemfile]: If we can replace cool.io with async gems, async-http should be default dependency.

Make sense 👍

**Pull request merged**